File: concept/type/impl/RelationTypeImpl.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.concept.type.impl;

import grakn.client.Grakn;
import grakn.client.concept.thing.impl.RelationImpl;
import grakn.client.concept.thing.impl.ThingImpl;
import grakn.client.concept.type.RelationType;
import grakn.protocol.ConceptProto;

import java.util.stream.Stream;

public class RelationTypeImpl extends ThingTypeImpl implements RelationType {

    RelationTypeImpl(String label, boolean isRoot) {
        super(label, isRoot);
    }

    public static RelationTypeImpl of(ConceptProto.Type typeProto) {
        return new RelationTypeImpl(typeProto.getLabel(), typeProto.getRoot());
    }

    @Override
    public RelationTypeImpl.Remote asRemote(Grakn.Transaction transaction) {
        return new RelationTypeImpl.Remote(transaction, getLabel(), isRoot());
    }

    @Override
    public RelationTypeImpl asRelationType() {
        return this;
    }

    public static class Remote extends ThingTypeImpl.Remote implements RelationType.Remote {

        public Remote(Grakn.Transaction transaction, String label, boolean isRoot) {
            super(transaction, label, isRoot);
        }

        @Override
        public RelationTypeImpl.Remote asRemote(Grakn.Transaction transaction) {
            return new RelationTypeImpl.Remote(transaction, getLabel(), isRoot());
        }

        @Override
        public final RelationImpl create() {
            final ConceptProto.Type.Req.Builder method = ConceptProto.Type.Req.newBuilder().setRelationTypeCreateReq(
                    ConceptProto.RelationType.Create.Req.getDefaultInstance());
            return RelationImpl.of(execute(method).getRelationTypeCreateRes().getRelation());
        }

        @Override
        public final RoleTypeImpl getRelates(String roleLabel) {
            final ConceptProto.Type.Req.Builder method = ConceptProto.Type.Req.newBuilder().setRelationTypeGetRelatesForRoleLabelReq(
                    ConceptProto.RelationType.GetRelatesForRoleLabel.Req.newBuilder().setLabel(roleLabel));
            final ConceptProto.RelationType.GetRelatesForRoleLabel.Res res = execute(method).getRelationTypeGetRelatesForRoleLabelRes();
            if (res.hasRoleType()) return TypeImpl.of(res.getRoleType()).asRoleType();
            else return null;
        }

        @Override
        public final Stream<RoleTypeImpl> getRelates() {
            return typeStream(
                    ConceptProto.Type.Req.newBuilder().setRelationTypeGetRelatesReq(
                            ConceptProto.RelationType.GetRelates.Req.getDefaultInstance()),
                    res -> res.getRelationTypeGetRelatesRes().getRolesList()
            ).map(TypeImpl::asRoleType);
        }

        @Override
        public final void setRelates(String roleLabel) {
            execute(ConceptProto.Type.Req.newBuilder().setRelationTypeSetRelatesReq(
                    ConceptProto.RelationType.SetRelates.Req.newBuilder().setLabel(roleLabel)));
        }

        @Override
        public final void setRelates(String roleLabel, String overriddenLabel) {
            execute(ConceptProto.Type.Req.newBuilder().setRelationTypeSetRelatesReq(
                    ConceptProto.RelationType.SetRelates.Req.newBuilder().setLabel(roleLabel).setOverriddenLabel(overriddenLabel)));
        }

        @Override
        public final void unsetRelates(String roleLabel) {
            execute(ConceptProto.Type.Req.newBuilder().setRelationTypeUnsetRelatesReq(
                    ConceptProto.RelationType.UnsetRelates.Req.newBuilder().setLabel(roleLabel)));
        }

        @Override
        public final void setSupertype(RelationType relationType) {
            super.setSupertype(relationType);
        }

        @Override
        public RelationTypeImpl getSupertype() {
            final ThingTypeImpl supertype = super.getSupertype();
            return supertype != null ? supertype.asRelationType() : null;
        }

        @Override
        public final Stream<RelationTypeImpl> getSubtypes() {
            return super.getSubtypes().map(ThingTypeImpl::asRelationType);
        }

        @Override
        public final Stream<RelationImpl> getInstances() {
            return super.getInstances().map(ThingImpl::asRelation);
        }

        @Override
        public RelationTypeImpl.Remote asRelationType() {
            return this;
        }
    }
}

File: concept/type/impl/RelationTypeImpl.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
    ThingTypeImpl,
    RemoteThingTypeImpl,
    RelationType,
    RemoteRelationType,
    Grakn,
    Stream,
    RelationImpl,
    RoleTypeImpl, TypeImpl,
} from "../../../dependencies_internal";
import Transaction = Grakn.Transaction;
import ConceptProto, { Type as TypeProto } from "grakn-protocol/protobuf/concept_pb";

export class RelationTypeImpl extends ThingTypeImpl implements RelationType {
    protected constructor(label: string, isRoot: boolean) {
        super(label, isRoot);
    }

    static of(typeProto: TypeProto): RelationTypeImpl {
        return new RelationTypeImpl(typeProto.getLabel(), typeProto.getRoot());
    }

    asRemote(transaction: Transaction): RemoteRelationTypeImpl {
        return new RemoteRelationTypeImpl(transaction, this.getLabel(), this.isRoot())
    }
}

export class RemoteRelationTypeImpl extends RemoteThingTypeImpl implements RemoteRelationType {
    constructor(transaction: Transaction, label: string, isRoot: boolean) {
        super(transaction, label, isRoot);
    }

    asRemote(transaction: Transaction): RemoteRelationTypeImpl {
        return new RemoteRelationTypeImpl(transaction, this.getLabel(), this.isRoot())
    }

    create(): Promise<RelationImpl> {
        const method = new ConceptProto.Type.Req().setRelationTypeCreateReq(new ConceptProto.RelationType.Create.Req());
        return this.execute(method).then(res => RelationImpl.of(res.getRelationTypeCreateRes().getRelation()));
    }

    getRelates(roleLabel: string): Promise<RoleTypeImpl>;
    getRelates(): Stream<RoleTypeImpl>;
    getRelates(roleLabel?: string): Promise<RoleTypeImpl> | Stream<RoleTypeImpl> {
        if (roleLabel != null) {
            const method = new ConceptProto.Type.Req().setRelationTypeGetRelatesForRoleLabelReq(
                new ConceptProto.RelationType.GetRelatesForRoleLabel.Req().setLabel(roleLabel));
            return this.execute(method).then(res => {
                const getRelatesRes = res.getRelationTypeGetRelatesForRoleLabelRes();
                if (getRelatesRes.hasRoleType()) return TypeImpl.of(getRelatesRes.getRoleType()) as RoleTypeImpl;
                else return null;
            });
        }

        return this.typeStream(
            new ConceptProto.Type.Req().setRelationTypeGetRelatesReq(new ConceptProto.RelationType.GetRelates.Req()),
            res => res.getRelationTypeGetRelatesRes().getRoleList()) as Stream<RoleTypeImpl>;
    }

    setRelates(roleLabel: string): Promise<void>;
    setRelates(roleLabel: string, overriddenLabel: string): Promise<void>;
    async setRelates(roleLabel: string, overriddenLabel?: string): Promise<void> {
        const setRelatesReq = new ConceptProto.RelationType.SetRelates.Req().setLabel(roleLabel);
        if (overriddenLabel != null) setRelatesReq.setOverriddenLabel(overriddenLabel);
        await this.execute(new ConceptProto.Type.Req().setRelationTypeSetRelatesReq(setRelatesReq));
    }

    async unsetRelates(roleLabel: string): Promise<void> {
        await this.execute(new ConceptProto.Type.Req()
            .setRelationTypeUnsetRelatesReq(new ConceptProto.RelationType.UnsetRelates.Req().setLabel(roleLabel)));
    }

    setSupertype(relationType: RelationType): Promise<void> {
        return super.setSupertype(relationType);
    }

    getSupertype(): Promise<RelationTypeImpl> {
        return super.getSupertype() as Promise<RelationTypeImpl>;
    }

    getSupertypes(): Stream<RelationTypeImpl> {
        return super.getSupertypes() as Stream<RelationTypeImpl>;
    }

    getSubtypes(): Stream<RelationTypeImpl> {
        return super.getSubtypes() as Stream<RelationTypeImpl>;
    }

    getInstances(): Stream<RelationImpl> {
        return super.getInstances() as Stream<RelationImpl>;
    }
}
